### PR TITLE
feature/eng-91: the long awaited refresh on Actions - Builder Pattern! 🎩

### PR DESCRIPTION
**Pull request by 4x8Matrix**
## Context
Addresses https://linear.app/metrik/issue/ENG-91/cleanup-implementation-of-sdk-actions

This impl doesn't actually register anything w the backend at the moment, it's just a refactor of the Actions - i'll be working on refactoring the SDKs comms ASAP!

### CODE REVIEW

**Comment by ethndotsh** [comment]
Waiting on possible enhanced typesafety and testing

**Comment by ethndotsh** [comment]
> Waiting on possible enhanced typesafety and testing

Also middlewares

File: Src/API/ActionBuilder.lua [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+function Action.Prototype:CanRun(...: unknown) -- todo: figure out how the 'middleware' impl is going to look.
+	return true
+end
+
+function Action.Prototype:OnRun(...: unknown)
+	print(`Method ':OnRun' has been called on Action<"{self.Name}">`)
+end
+
+function ActionBuilder.Prototype.SetName(self: ActionBuilder, actionName: string): ActionBuilder
```
Comment (ethndotsh) [inline Src/API/ActionBuilder.lua:40]: We should set a max length limit for the name, I forgot what it's set to in the API (or if it's even there.)

File: Tests/Development.server.lua [hunk 2]
```diff
@@ -7,7 +7,25 @@ MetrikSDK.Server:SetProjectId(ServerStorage.__Project_Id.Value)
...
+		:SetName("Hello, world!")
+		:SetDescription("A simple 'Hello, World' print!")
+		:AddArgument("String", {
+			Type = "string",
+			Required = false,
+			Description = "Anything else you want to print?"
+		})
+		:Build()
+
+	function action:OnRun(source: string?)
```
Comment (ethndotsh) [inline Tests/Development.server.lua:22]: would it make sense to change this to an event of some sort and add `:Connect`? Doesn't that match existing Roblox patterns

File: Tests/Development.server.lua [hunk 3]
```diff
@@ -7,7 +7,25 @@ MetrikSDK.Server:SetProjectId(ServerStorage.__Project_Id.Value)
 MetrikSDK.Server:SetAuthenticationToken(ServerStorage.__Project_Auth.Value)
 
 MetrikSDK.Server:InitializeAsync():andThen(function()
-	-- warn("Metrik SDK loaded!")
+	warn("Metrik SDK loaded!")
+	
+	local action = MetrikSDK.Server.ActionBuilder.new()
+		:SetName("Hello, world!")
+		:SetDescription("A simple 'Hello, World' print!")
+		:AddArgument("String", {
```
Comment (ethndotsh) [inline Tests/Development.server.lua:15]: What about a shorthand/bulk argument 'adder' as `:AddArguments`?

File: Src/API/ActionBuilder.lua (hunk 1 above)
Comment (4x8Matrix) [reply Src/API/ActionBuilder.lua:40]: there wasn't a limit :smile:

File: Tests/Development.server.lua (hunk 2 above)
Comment (4x8Matrix) [reply Tests/Development.server.lua:22]: the Existing roblox pattern is mostly singletons - this is an object (for reference, you could take a look at a DataStorePages instance) and so we're probably not going to provide signals, but may be something to do later down the line if requested.

File: Tests/Development.server.lua (hunk 3 above)
Comment (4x8Matrix) [reply Tests/Development.server.lua:15]: Probably worth keeping it simple atm, then scaling it up once we know things are up n working :smile:

File: Src/API/ActionBuilder.lua (hunk 1 above)
Comment (ethndotsh) [reply Src/API/ActionBuilder.lua:40]: set it at like 50 chars and forget it imo

File: Tests/Development.server.lua (hunk 3 above)
Comment (ethndotsh) [reply Tests/Development.server.lua:15]: sure

File: Tests/Development.server.lua (hunk 2 above)
Comment (ethndotsh) [reply Tests/Development.server.lua:22]: ok

**Comment by ethndotsh** [comment]
Will merge after we look more into enhanced typesafety

**Pull request merged**